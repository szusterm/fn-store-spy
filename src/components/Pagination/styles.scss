@import '../../scss/variables';

.pagination {
	&--button {
		$button-height: 50px;
		$normal-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);
		$hover-shadow: 0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.23);

		cursor: pointer;

		width: 100%;
		height: $button-height;
		padding: 10px;
		border-width: 0;
		background-color: $body-bg;
		color: $body-color;

		display: flex;
		justify-content: center;
		align-items: center;

		transition-duration: 100ms;
		box-shadow: $normal-shadow;

		&:hover {
			box-shadow: $hover-shadow
		}

		&:disabled {
			cursor: default;
			background-color: rgba(0, 0, 0, 0.1);
			box-shadow: $normal-shadow;
		}

		&--previous {
			border-top-left-radius: $button-height/2;
			border-bottom-left-radius: $button-height/2;
		}

		&--next {
			border-top-right-radius: $button-height/2;
			border-bottom-right-radius: $button-height/2;
		}
	}
}