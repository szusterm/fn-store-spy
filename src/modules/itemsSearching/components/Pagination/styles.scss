@import '~material-colors/dist/colors';

@import '~scss/_shadow';
@import '~scss/_transition';
@import '~scss/_opacity-colors';
@import '~scss/_input';
@import '~scss/_center-items';

.pagination {
	&--button {
		@include default-input();
		$button-height: 50px;

		cursor: pointer;

		width: 100%;
		height: $button-height;
		padding: 10px;
		border-width: 0;
		color: $md-grey-900;

		@include center-items();
		@include transition(sh);
		@include shadow(sm);
		@include white-background(0.9);

		&:hover, &:focus {
			@include shadow(md);
			@include white-background(1);
		}

		&:disabled {
			cursor: default;
			@include shadow(sm);
			@include white-background(0.3);
		}

		&--previous {
			border-top-right-radius: 0;
			border-bottom-right-radius: 0;
		}

		&--next {
			border-top-left-radius: 0;
			border-bottom-left-radius: 0;
		}
	}
}