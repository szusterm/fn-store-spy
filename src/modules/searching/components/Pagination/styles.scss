@import '~scss/_variables';
@import '~scss/_shadow';
@import '~scss/_transition';
@import '~scss/_opacity-colors';
@import '~scss/_input-reset';
@import '~scss/_center-items';

.pagination {
	&--button {
		@include input-reset();
		$button-height: 50px;

		cursor: pointer;

		width: 100%;
		height: $button-height;
		padding: 10px;
		border-width: 0;
		color: $body-color;

		@include center-items();
		@include transition(sh);
		@include shadow(sm);
		@include white-background(0.9);

		&:hover, &:focus {
			@include shadow(md);
			@include white-background(1);
		}

		&:disabled {
			cursor: default;
			@include shadow(sm);
			@include white-background(0.3);
		}

		&--previous {
			border-top-left-radius: $button-height/2;
			border-bottom-left-radius: $button-height/2;
		}

		&--next {
			border-top-right-radius: $button-height/2;
			border-bottom-right-radius: $button-height/2;
		}
	}
}