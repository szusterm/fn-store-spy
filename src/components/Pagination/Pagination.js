import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import './styles.scss';

class Pagination extends Component {
	constructor() {
		super();

		this.state = {
			page: 1
		};

		this.previousPage = this.previousPage.bind(this);
		this.nextPage = this.nextPage.bind(this);
	}

	updatePage(newPage = 1) {
		const newState = this.state;
		newState.page = newPage;
		this.setState(newState);

		this.props.onChange(newState.page);
	}

	nextPage() {
		const {page} = this.state;
		const newPage = page + 1;
		this.updatePage(newPage);
	}

	previousPage() {
		const {page} = this.state;

		if (page > 1) {
			const newPage = page - 1;
			this.updatePage(newPage);
		}
	}

	render() {
		return (
			<div className={'container-fluid pagination'}>
				<div className={'row'}>
					<div className={'col-6'}>
						<button
							className={'pagination--button pagination--button--previous'}
							disabled={(this.state.page <= 1)}
							onClick={this.previousPage}
						>
							Previous
						</button>
					</div>
					<div className={'col-6'}>
						<button
							className={'pagination--button pagination--button--next'}
							disabled={!this.props.nextPageAvailable}
							onClick={this.nextPage}
						>
							Next
						</button>
					</div>
				</div>
			</div>
		);
	}
}

Pagination.propTypes = {
	onChange: PropTypes.func,
	nextPageAvailable: PropTypes.bool //redux
};

const mapStateToProps = (state) => {
	const {nextPageAvailable} = state;
	return {nextPageAvailable};
};

export default connect(mapStateToProps, null)(Pagination);