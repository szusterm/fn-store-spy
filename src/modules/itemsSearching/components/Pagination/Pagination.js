import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import {setPageFilter} from '../../actions';

import './styles.scss';

export class Pagination extends Component {
	constructor() {
		super();

		this.setPreviousPage = this.setPreviousPage.bind(this);
		this.setNextPage = this.setNextPage.bind(this);
	}

	setNextPage() {
		const {page} = this.props;
		const newPage = page + 1;
		this.props.setPageFilter(newPage);

		this.props.onChange();
	}

	setPreviousPage() {
		const {page} = this.props;

		if (page > 1) {
			const newPage = page - 1;
			this.props.setPageFilter(newPage);
		}

		this.props.onChange();
	}

	render() {
		const {page, nextPageAvailable} = this.props;

		return (
			<div className={'col-12 col-sm-10 col-md-8 col-lg-6 col-xl-4'}>
				<div className={'container-fluid pagination'}>
					<div className={'row'}>
						<div className={'col-6'}>
							<button
								className={'pagination--button pagination--button--previous'}
								disabled={(page <= 1)}
								onClick={this.setPreviousPage}
							>
								Previous
							</button>
						</div>
						<div className={'col-6'}>
							<button
								className={'pagination--button pagination--button--next'}
								disabled={!nextPageAvailable}
								onClick={this.setNextPage}
							>
								Next
							</button>
						</div>
					</div>
				</div>
			</div>
		);
	}
}

Pagination.propTypes = {
	onChange: PropTypes.func,
	setPageFilter: PropTypes.func, //redux
	page: PropTypes.number, //redux
	nextPageAvailable: PropTypes.bool //redux
};

const mapStateToProps = (state) => {
	const {nextPageAvailable} = state.itemsSearching;
	const {page} = state.itemsSearching.filters;

	return {nextPageAvailable, page};
};

const mapDispatchToProps = {setPageFilter};

export default connect(mapStateToProps, mapDispatchToProps)(Pagination);