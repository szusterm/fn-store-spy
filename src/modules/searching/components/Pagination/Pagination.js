import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import {setPageFilter} from '../../../../redux/actions';

import './styles.scss';

export class Pagination extends Component {
	constructor() {
		super();

		this.previousPage = this.previousPage.bind(this);
		this.nextPage = this.nextPage.bind(this);
	}

	nextPage() {
		const {page} = this.props;
		const newPage = page + 1;
		this.props.setPageFilter(newPage);

		this.props.onChange();
	}

	previousPage() {
		const {page} = this.props;

		if (page > 1) {
			const newPage = page - 1;
			this.props.setPageFilter(newPage);
		}

		this.props.onChange();
	}

	render() {
		const {page, nextPageAvailable} = this.props;

		return (
			<div className={'container-fluid pagination'}>
				<div className={'row'}>
					<div className={'col-6'}>
						<button
							className={'pagination--button pagination--button--previous'}
							disabled={(page <= 1)}
							onClick={this.previousPage}
						>
							Previous
						</button>
					</div>
					<div className={'col-6'}>
						<button
							className={'pagination--button pagination--button--next'}
							disabled={!nextPageAvailable}
							onClick={this.nextPage}
						>
							Next
						</button>
					</div>
				</div>
			</div>
		);
	}
}

Pagination.propTypes = {
	onChange: PropTypes.func,
	setPageFilter: PropTypes.func, //redux
	page: PropTypes.number, //redux
	nextPageAvailable: PropTypes.bool //redux
};

const mapStateToProps = (state) => {
	const {nextPageAvailable} = state.searching;
	const {page} = state.searching.filters;

	return {nextPageAvailable, page};
};

const mapDispatchToProps = {setPageFilter};

export default connect(mapStateToProps, mapDispatchToProps)(Pagination);