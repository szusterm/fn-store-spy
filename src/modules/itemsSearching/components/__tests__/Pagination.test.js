import React from 'react';
import {shallow} from 'enzyme';

import {Pagination} from '../Pagination/Pagination';

const setup = (propOverrides) => {
	const props = Object.assign({
		onChange: () => true,
		setPageFilter: () => true, //redux
		page: 1, //redux
		nextPageAvailable: true //redux
	}, propOverrides);

	const wrapper = shallow(<Pagination {...props}/>);

	return {props, wrapper};
};

describe('Pagination Component', () => {
	it('renders correctly', () => {
		const {wrapper} = setup();

		expect(wrapper).toMatchSnapshot();
	});

	it('calls onChange() after click the next or the previous button', () => {
		const startPage = 2;

		const {wrapper, props} = setup({
			onChange: jest.fn(),
			page: startPage,
			nextPageAvailable: true
		});

		wrapper.find('.pagination--button--previous').simulate('click');
		wrapper.find('.pagination--button--next').simulate('click');

		expect(props.onChange).toHaveBeenCalledTimes(2);
	});

	it('calls a redux action to change the page after click a button', () => {
		const startPage = 2;

		const {wrapper, props} = setup({
			setPageFilter: jest.fn()
		});

		wrapper.setProps({page: startPage});
		wrapper.find('.pagination--button--previous').simulate('click');
		expect(props.setPageFilter).toHaveBeenCalledWith(startPage - 1);

		wrapper.setProps({page: startPage});
		wrapper.find('.pagination--button--next').simulate('click');
		expect(props.setPageFilter).toHaveBeenCalledWith(startPage + 1);
	});

	it('makes the previous button disabled, if the page equals to 1', () => {
		const {wrapper} = setup();

		wrapper.setProps({page: 1});
		const isDisabledWithLower = wrapper.find('.pagination--button--previous').props().disabled;
		expect(isDisabledWithLower).toBe(true);

		wrapper.setProps({page: 2});
		const isDisabledWithHigher = wrapper.find('.pagination--button--previous').props().disabled;
		expect(isDisabledWithHigher).toBe(false);
	});

	it('makes the next button disabled, if the next page does not exists', () => {
		const {wrapper} = setup();

		wrapper.setProps({nextPageAvailable: false});
		const isDisabledWithoutNextPage = wrapper.find('.pagination--button--next').props().disabled;
		expect(isDisabledWithoutNextPage).toBe(true);

		wrapper.setProps({nextPageAvailable: true});
		const isDisabledWithNextPage = wrapper.find('.pagination--button--next').props().disabled;
		expect(isDisabledWithNextPage).toBe(false);
	});
});