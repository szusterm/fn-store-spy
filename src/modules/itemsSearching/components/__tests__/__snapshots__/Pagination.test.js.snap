// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Component renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Pagination
    nextPageAvailable={true}
    onChange={[Function]}
    page={1}
    setPageFilter={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="pagination"
      >
        <button
          className="pagination--button pagination--button--previous"
          disabled={true}
          onClick={[Function]}
        >
          Previous
        </button>
        <button
          className="pagination--button pagination--button--next"
          disabled={false}
          onClick={[Function]}
        >
          Next
        </button>
      </div>,
      "className": "col-12 col-sm-10 col-md-8 col-lg-6 col-xl-4",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="pagination--button pagination--button--previous"
            disabled={true}
            onClick={[Function]}
          >
            Previous
          </button>,
          <button
            className="pagination--button pagination--button--next"
            disabled={false}
            onClick={[Function]}
          >
            Next
          </button>,
        ],
        "className": "pagination",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Previous",
            "className": "pagination--button pagination--button--previous",
            "disabled": true,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Previous",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Next",
            "className": "pagination--button pagination--button--next",
            "disabled": false,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Next",
          "type": "button",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="pagination"
        >
          <button
            className="pagination--button pagination--button--previous"
            disabled={true}
            onClick={[Function]}
          >
            Previous
          </button>
          <button
            className="pagination--button pagination--button--next"
            disabled={false}
            onClick={[Function]}
          >
            Next
          </button>
        </div>,
        "className": "col-12 col-sm-10 col-md-8 col-lg-6 col-xl-4",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="pagination--button pagination--button--previous"
              disabled={true}
              onClick={[Function]}
            >
              Previous
            </button>,
            <button
              className="pagination--button pagination--button--next"
              disabled={false}
              onClick={[Function]}
            >
              Next
            </button>,
          ],
          "className": "pagination",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Previous",
              "className": "pagination--button pagination--button--previous",
              "disabled": true,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Previous",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Next",
              "className": "pagination--button pagination--button--next",
              "disabled": false,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Next",
            "type": "button",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
